src/software/amazon/ionschema/internal/util/Range.kt

```diff
@@ -0,0 +1,62 @@
+package software.amazon.ionschema.internal.util
+
+import software.amazon.ion.IonList
+import software.amazon.ion.IonSymbol
+import software.amazon.ion.IonValue
+import software.amazon.ionschema.InvalidSchemaException
+
+/**
+ * Enum representing the support types of ranges.
+ */
+internal enum class RangeType {
+    INT,
+    INT_NON_NEGATIVE,
+    ION_NUMBER,
+    ION_TIMESTAMP,
+    ION_TIMESTAMP_PRECISION,
+}
+
+/**
+ * Interface for all range implementations.
+ */
+internal interface Range<in T : Any> {
+    fun contains(value: T): Boolean
+}
+
+/**
+ * Factory method for instantiating Range<T> objects.
+ */
+internal class RangeFactory {
+    companion object {
+        @JvmStatic
+        fun <T : Any> rangeOf(ion: IonValue, rangeType: RangeType): Range<T> {
+            if (ion.isNullValue) {
+                throw InvalidSchemaException("Invalid range $ion")
+            }
+
+            val ionList = if (ion !is IonList) {
+                    val range = ion.system.newList(ion.clone(), ion.clone())
+                    range.addTypeAnnotation("range")
+                    range
+                } else {
+                    ion
+                }
+
+            if (ionList.size != 2 || ionList[0].isNullValue || ionList[1].isNullValue
+                    || ((ionList[0] as? IonSymbol)?.stringValue() == "min"
+                            && ((ionList[1] as? IonSymbol)?.stringValue() == "max"))) {
```

Comment:
Am I reading this correctly?  `range::[min, max]` is considered an invalid range?  It seems you also have asserting this to be invalid, however the specification has examples under the "Ranges" heading [here](https://amzn.github.io/ion-schema/docs/spec.html#constraints) that seem to suggest it should be valid.

Reply:
I looked some more and I found tests where `range::[min, max]` was considered valid and from a moderately quick read of the specification I don't see under what circumstances it would be valid or not...